_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,598 @@
+        IDT_LIST:          56 B         2 KB      2.73%
+[...]
+[188/188] Linking C executable zephyr/zephyr.elf
+```
+
+Zephyr already has some code size statistics at the end of their build, which is
+great! However, there is room for improvement.
+
+To help generate data for me to work with, I spun up an AWS Spot EC2 instance
+and built the most recent 2,500 commits of Zephyr from the master branch as of
+March 14th. I would have built more but there was an issue with the CMake build
+system and dependencies unfortunately.
+
+The process for this was quite simple. Using `git rev-list --first-parent`, we
+can get all the revisions that were committed into the master branch
+chronological order.
+
+```
+$ git rev-list --first-parent master
+22b9167acb52386c13be6d22223fd3d585d57a3d
+7cb8f5c61e6081475223586349b88c5a34eb7f38
+[...]
+```
+
+Once we have all the commits, we can run a Python script that builds every
+commit.
+
+```python
+for commit in commits:
+    try:
+        os.system("git checkout {}".format(commit))
+        os.system("west update")
+        os.system("west build -b cc3220sf_launchxl samples/net/lwm2m_client --pristine")
+        # Here, we calculate and update the code size.
+    except:
+        continue
+```
+
+> To whomever it concerns: I _love_ the `--pristine` option in the `west` tool.
+
+### Capturing the Sizes
+
+To get a generic code size calculation, we can use `size`, or in the case of
+this blog post, `arm-none-eabi-size`. This captures the `text`, `data`, and
+`bss` region code sizes and prints them to the terminal.
+
+```
+$ arm-none-eabi-size zephyr/zephyr.elf
+   text    data     bss     dec     hex filename
+ 130780    3356   21167  155303   25ea7 build/zephyr/zephyr.elf
+```
+
+There are a number of ways to capture this data, and some projects or teams want
+to track other data important to them. These might include specific linker
+region sizes, resource pack size (fonts, images, etc), or possibly the largest
+10 files or symbols in the codebase. Whatever is important to you and your team,
+track it!
+
+### Organizing the Sizes
+
+Our goal is to keep track of the sizes of the `text`, `data`, and `bss`
+**regions** for every **commit** and every **build**. By storing the commit
+revision and the build type, we can perform code size deltas between any
+revisions built with the same configuration.
+
+To ensure that we can calculate the deltas between commits properly, we also
+need to keep track of the "parent" commit, since the commit that comes before it
+in the `git log` is not necessarily the parent.
+
+A commit's parent can be found using the following git command:
+
+```
+$ git rev-parse a2bc514653^1
+c1fdf98ba5eb393a7646775921630bf4323ed832
+```
+
+To summarize, for every commit, we will store the following information for each
+**commit and build** combination:
+
+- `revision`: `a2bc51465395057e3830614b20ceea10976204bb`
+- `parent_revision`: `c1fdf98ba5eb393a7646775921630bf4323ed832`
```

Comment:
Don't forget to add commit date here, if you end up adding it to your code

Reply:
Done!